src/python/pants/engine/target.py

```diff
@@ -1052,6 +1055,45 @@ def validate(self) -> None:
                 "`TargetGenerator.moved_field`s, to avoid redundant graph edges."
             )
 
+    @classmethod
+    def register_plugin_field(cls, field: Type[Field], *, as_moved_field=False) -> UnionRule:
+        if as_moved_field:
+            return UnionRule(cls.MovedPluginField, field)
+        else:
+            return super().register_plugin_field(field)
+
+    @classmethod
+    @memoized_method
+    def _find_plugin_fields(cls, union_membership: UnionMembership) -> tuple[type[Field], ...]:
+        return (
+            *cls._find_copied_plugin_fields(union_membership),
+            *cls._find_moved_plugin_fields(union_membership),
+        )
+
+    @final
+    @classmethod
+    @memoized_method
+    def _find_moved_plugin_fields(
+        cls, union_membership: UnionMembership
+    ) -> tuple[type[Field], ...]:
+        result: set[type[Field]] = set()
+        classes = [cls]
+        while classes:
+            cls = classes.pop()
+            classes.extend(cls.__bases__)
+            if issubclass(cls, TargetGenerator):
+                result.update(cast("set[type[Field]]", union_membership.get(cls.MovedPluginField)))
+
+        return tuple(result)
```

Comment:
We should sort, or do some other way of ensuring the return is deterministic, since `set`s don't have deterministic order across runs (not technically a problem here, since this would only get memoized at the process level anyways, but better be safe and consistent)

Reply:
The impl structure is taken from `Target._find_plugin_fields` which does not have the sort:
```python
    @final
    @classmethod
    @memoized_method
    def _find_plugin_fields(cls, union_membership: UnionMembership) -> tuple[type[Field], ...]:
        result: set[type[Field]] = set()
        classes = [cls]
        while classes:
            cls = classes.pop()
            classes.extend(cls.__bases__)
            if issubclass(cls, Target):
                result.update(cast("set[type[Field]]", union_membership.get(cls.PluginField)))

        return tuple(result)
```

Should we change both? @kaos and I also weighed the balance on whether we should attempt to DRY.

Reply:
good point Josh. I think the question to ask is, if this returned value could end up in a datastructure returned from a rule..? in which case we certainly do want it to be stable also across sessions.
For that reason, I'm +1 to sort it, to be safe. (it may not be the case today, but a future me would certainly overlook this if I were to introduce that.)

Reply:
@kaos @thejcannon do we have precedent for sorting a list of classes? Would we do it by some kind of fully qualified name to be stable? types otherwise have no built-in ordering (at least not in py3!).

Reply:
huh.. not sure. maybe punt on this then. I was going to say that we could sort based on the `str` or `repr` representation of the type, but that starts to feel like a hack..

Reply:
We don't have to sort it. Just make sure the order is deterministic. That could mean using an OrderedSet

Reply:
I fail to see what difference the ordered set would do here. As you said, the result is memoized, so will only be evaluated once per run, and between runs if the order of types found by the union memberships shift, the result will reflect that shift also with the OrderedSet, right?

Reply:
It's mostly a consistency thing. Most code in this codebase is subject to bugs if they use set, so generally speaking we shouldn't be using sets.

Specifically this one is probably ok, for the reasons we've stated (hence approving the PR). (I say probably because there's a chance the memoization cache gets cleared and therefore order does matter. I know at work we clear our caches)

Reply:
well, worst case is invalidating cache entries down the line that could've been avoided, when someone clears the caches. That feels like not a big deal tbh :)
(and we do return a tuple, so the use of the set is very limited)

Reply:
I think we agree this is good enough for now. (it's not new after all, so no worse off than before.)

Reply:
🤦 its only now that it is merged that it occurs to me ...

these are the Field types of a Target, not any classes. We could have sorted them by their `.alias` which is guaranteed to exist _and_ should be unique for a single target.

Reply:
Feel free to submit a refactor PR if you think the improvement is worth it (no harm done for not realising earlier!).

Reply:
good point. yea we're a whole bunch here that didn't think of it, so a round of 🤦🏽  to go around 😛 

I think this is mostly hypothetical any way, but could be worth fixing regardless.. (for the odd edge cases etc)